dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,443 @@
 	//
-	// +required
-	Class string `json:"class"`
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,1,opt,name=config"`
 
 	// AdminAccess indicates that this is a claim for administrative access
-	// to the devices. Claims with AdminAccess are expected to be used for
+	// to the device(s). Claims with AdminAccess are expected to be used for
 	// monitoring or other management services for a device.  They ignore
 	// all ordinary claims to the device with respect to access modes and
-	// any resource allocations. Ability to create these claims is
-	// controlled via ResourceQuota.
+	// any resource allocations. Ability to request this kind of access is
+	// controlled via ResourceQuota in the resource.k8s.io API.
 	//
-	// Default is false. If true, a DeviceClass must be specified so that
-	// the Driver is known, and only the Constraints fields will be taken
-	// into account for device selection. All devices meeting the
-	// Constraints expressions will be made available as part of the claim
-	// and may be assigned to a container.
+	// Can be combined with a range to ask for access to all devices
+	// on a node which match the requrirements.
 	//
-	// NOTE: This cannot appear in class because the quota code does not do
-	// an indirection. Still searching for a better way to handle this use
-	// case, without creating a new top-level type for it.
+	// Default is false.
 	//
 	// +optional
 	AdminAccess *bool `json:"adminAccess,omitempty"`
 
-	// Constraints is a CEL expression that operates on device attributes.
-	// In order for a device to be considered, this CEL expression and the
-	// Constraints expression from the DeviceClass must both be true.
+	// FUTURE EXTENSION:
+	// - Count with min, max (including "all that are available").
+	// - Constraints for the set of devices belonging to this request.
+
+	// Requirements describe additional contraints that all must be met by a device
+	// to satisfy the request.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
-
-	// Device claims may be satisfied by choosing multiple devices instead
-	// of just a single device.
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
 
-	// MaxDevices allows the user to specify a maximum number of devices
-	// that may be allocated to sastisfy this claim. The default is equal
-	// to requests.devices. Thus, without specifying MaxDevices, the user
-	// will get exactly the number of devices specified in
-	// requests.devices.
-	//
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+	// Allocation is set once the claim has been allocated successfully.
 	// +optional
-	MaxDevices *int `json:"maxDevices,omitempty"`
+	Allocation *AllocationResult `json:"allocation,omitempty" protobuf:"bytes,2,opt,name=allocation"`
```

Comment:
This structure doesn't quite work now, with requests nested in a claim. We need an allocation result per request. Or some other structure that allows us to connect allocations back to the specific requests that resulted in them, especially since requests in the same claim could be satisfied by different drivers.

I think this can go on the TO DO list too. One advantage of the WG repo is that we can merge stuff even if it's not really done. This PR is killing my browser!

Reply:
That is what `RequestAllocationResult.RequestName` is for. If the user selects a specific request for a container via that name, then kubelet can look up the right result.